C2S Packets validation

sruon opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Based on a recent discussion, I am proposing that LSB:
- Aligns inbound packets definitions on XiPackets documentation
- Perform sanity checks on packets before processing them

Intents:
- Stop blindly trusting received data
- Get rid of `data.ref<uint8>(0x0A)` calls
- Allow cross-referencing XiPackets documentation in an easier fashion

Flow:
- Packet parser goes through a new handler for these packets
- This handler enforces a 2-step process for each packet:
  - validate(): A set of rules that determines if a packet is legitimate, with built-in validators for common scenarios and a fallback to lambdas if needed
  - process(): What used to be in packet_system.cpp
- Packets that do not pass validation are not processed, and are logged.

Looking for feedback on design and/or ideas for more idiomatic C++ patterns. 
I have tried to make the definition of structures and their validation as "natural" as possible, with some complexity hidden behind macros and templates.
Not looking to solve every single complex validation scenarios but to at least have basic sanity checks.

## Steps to test these changes

<!-- Clear and detailed steps to test your changes here -->

sruon committed:
C2S Packets validation

Co-Authored-By: atom0s <atom0s@users.noreply.github.com>

zach2good commented on src/map/packet_system.cpp:
```diff
@@ -7564,25 +7416,21 @@ void SmallPacket0x11B(MapSession* const PSession, CCharEntity* const PChar, CBas
+    const T* packet = data.as<T>();
 
-    PChar->loc.zone->PushPacket(PChar, CHAR_INRANGE_SELF, std::make_unique<CCharEmotionJumpPacket>(PChar, targetIndex, extra));
+    if (const packet_validation_t result = packet->validate(PSession, PChar))
```
This would read nicer as:
```cpp
if (const auto result = packet->validate(PSession, PChar); result.valid())
{
    // valid path
}
else // invalid path
{
    ShowWarningFmt("Invalid {} packet from {}: {} ", packet->getName(), PChar->name, result.errorString());
}
```

zach2good commented on src/map/packets/c2s/validation.h:
```diff
@@ -0,0 +1,143 @@
+
+#pragma once
+
+struct packet_validation_t
```
Can you make this a class and encapsulate the member access? I'd probably also remove the `bool()` operator, in favor of `valid()` and `setValid(false)` etc.

Maybe also name it `PacketValidationResult`

zach2good commented on src/map/packets/c2s/validation.h:
```diff
@@ -0,0 +1,143 @@
+
+private:
+    packet_validation_t result_;
+};
```
Through all of these, whenever you use `auto`, try and make it `const auto` - since everything you're doing here is very much in the hot path, try and be as `const` as possible to give the compiler the best chance possible of generating the fastest code it can

zach2good commented on src/map/CMakeLists.txt:
```diff
@@ -4,6 +4,7 @@ add_subdirectory(items)
 add_subdirectory(los)
 add_subdirectory(lua)
 add_subdirectory(packets)
+add_subdirectory(packets/c2s)
```
At some point `packets/` should have its own `CMakeLists.txt` that then has `add_subdirectory(c2s)` and `add_subdirectory(s2c)`, but not needed in this PR

zach2good reviewed (approved):
Still great, just a couple of notes then ready to merge 👍

sruon commented:
Addressed the comments. Couple of comments from me:

- s2c packets will undergo the same treatment when done with c2s. CMakeLists will be updated at that time to be properly nested
- There is some degree of overlap with PacketGuard. I think we can merge the rate limiting and state checks in the validator. Not urgent, I'll keep thinking about it.
- Some packets are not named in XiPackets, I've taken the liberty of assigning them meaningful names but maybe we'd want to reflect that in their name and possibly typedef an aliased name

zach2good commented:
I suspect we can roll the rate limiting and packet ordering into this new system eventually, and then the state based validation can be it's own player-based thing
